DataDog/dd-trace-rb
# Add remote configuration dispatcher

**What does this PR do?**

Add a dispatcher.

**Motivation**

Remote configuration

**Additional Notes**

The dispatcher gets registered against using a receiver for a block to be called and a matcher to select relevant changes from the changeset.

Upon changes, the dispatcher matches changes against the matcher and calls the receiver, passing the subset of changes as well as the repository itself. This is because:

- A receiver might need not just the change but complement that change with the full state. therefore the repository state is needed.
- A receiver might need to understand a specific change in the changeset and act in a specific way according to the type (it might want to operate differently on an update vs a deletion) or content (it might want to look at the diff in state) so passing the current repository state is not sufficient. Therefore passing the list of changes is needed.

A generic matcher that takes a block is available, for other matchers to be implemented. A simple product matcher is implemented, based on the generic one.

**How to test the change?**

```
  require 'ddtrace'

  Datadog.configure do |c|
    c.agent.host = '192.168.135.133'
    c.diagnostics.debug = true
  end

  Datadog::Core::Remote.active_remote.sync
  repository = Datadog::Core::Remote.active_remote.client.instance_eval { @repository }

  require 'irb'

  sleep 1000
```

Wait a bit to receive first batch of RC. This should happen:

```
D, [2023-03-21T10:18:10.338264 #29926] DEBUG -- ddtrace: [ddtrace] (/Users/lloeki/Source/github.com/DataDog/dd-trace-rb-alt/lib/datadog/core/remote/client.rb:187:in `block (2 levels) in register_receivers') remote config change: #<Datadog::Core::Remote::Configuration::Path:0x00000001066225f8 @source="datadog", @org_id=603646, @product="ASM_DATA", @config_id="blocked_users", @name="config">
D, [2023-03-21T10:18:10.338312 #29926] DEBUG -- ddtrace: [ddtrace] (/Users/lloeki/Source/github.com/DataDog/dd-trace-rb-alt/lib/datadog/core/remote/client.rb:187:in `block (2 levels) in register_receivers') remote config change: #<Datadog::Core::Remote::Configuration::Path:0x00000001066223c8 @source="employee", @org_id=nil, @product="ASM_DD", @config_id="16.recommended.json", @name="config">
D, [2023-03-21T10:18:10.338326 #29926] DEBUG -- ddtrace: [ddtrace] (/Users/lloeki/Source/github.com/DataDog/dd-trace-rb-alt/lib/datadog/core/remote/client.rb:187:in `block (2 levels) in register_receivers') remote config change: #<Datadog::Core::Remote::Configuration::Path:0x0000000106622148 @source="datadog", @org_id=603646, @product="ASM", @config_id="blocking", @name="config">
D, [2023-03-21T10:18:10.338392 #29926] DEBUG -- ddtrace: [ddtrace] (/Users/lloeki/Source/github.com/DataDog/dd-trace-rb-alt/lib/datadog/core/remote/client.rb:187:in `block (2 levels) in register_receivers') remote config change: #<Datadog::Core::Remote::Configuration::Path:0x0000000106621950 @source="datadog", @org_id=603646, @product="ASM", @config_id="exclusion_filters", @name="config">
D, [2023-03-21T10:18:10.338406 #29926] DEBUG -- ddtrace: [ddtrace] (/Users/lloeki/Source/github.com/DataDog/dd-trace-rb-alt/lib/datadog/core/remote/client.rb:187:in `block (2 levels) in register_receivers') remote config change: #<Datadog::Core::Remote::Configuration::Path:0x00000001066202d0 @source="datadog", @org_id=603646, @product="ASM_DATA", @config_id="blocked_ips", @name="config">
```

Go in the DD [Application Security UI](https://app.datadoghq.com/security/appsec/denylist) and add a new IP to the list. Wait a bit. This should pop up:

```
D, [2023-03-21T10:20:50.616185 #29926] DEBUG -- ddtrace: [ddtrace] (/Users/lloeki/Source/github.com/DataDog/dd-trace-rb-alt/lib/datadog/core/remote/client.rb:187:in `block (2 levels) in register_receivers') remote config change: #<Datadog::Core::Remote::Configuration::Path:0x0000000106b17f68 @source="datadog", @org_id=603646, @product="ASM_DATA", @config_id="blocked_ips", @name="config">
```
